pdns/dnsdist-lua-bindings.cc

```diff
@@ -713,4 +714,97 @@ void setupLuaBindings(bool client)
     }
     return values;
   });
+
+  /* Key Value Store objects */
+  g_lua.writeFunction("KeyValueLookupKeySourceIP", []() {
+    return std::shared_ptr<KeyValueLookupKey>(new KeyValueLookupKeySourceIP());
+  });
+  g_lua.writeFunction("KeyValueLookupKeyQName", []() {
+    return std::shared_ptr<KeyValueLookupKey>(new KeyValueLookupKeyQName());
+  });
+  g_lua.writeFunction("KeyValueLookupKeySuffix", []() {
+    return std::shared_ptr<KeyValueLookupKey>(new KeyValueLookupKeySuffix());
+  });
+  g_lua.writeFunction("KeyValueLookupKeyTag", [](const std::string& tag) {
+    return std::shared_ptr<KeyValueLookupKey>(new KeyValueLookupKeyTag(tag));
+  });
+
+#ifdef HAVE_LMDB
+  g_lua.writeFunction("newLMDBKVStore", [client](const std::string& fname, const std::string& dbName) {
+    if (client) {
+      return std::shared_ptr<KeyValueStore>(nullptr);
+    }
+    return std::shared_ptr<KeyValueStore>(new LMDBKVStore(fname, dbName));
+  });
+#endif /* HAVE_LMDB */
+
+#ifdef HAVE_CDB
+  g_lua.writeFunction("newCDBKVStore", [client](const std::string& fname, time_t refreshDelay) {
+    if (client) {
+      return std::shared_ptr<KeyValueStore>(nullptr);
+    }
+    return std::shared_ptr<KeyValueStore>(new CDBKVStore(fname, refreshDelay));
+  });
+#endif /* HAVE_CDB */
+
+  g_lua.registerFunction<std::string(std::shared_ptr<KeyValueStore>::*)(const std::string&)>("lookup", [](std::shared_ptr<KeyValueStore>& kvs, const std::string& keyStr) {
```

Comment:
Couldn't we expose multiple `lookup(ComboAddress)`, `lookup(DNSName)` and `lookup(string)` ?

Reply:
LuaWrapper doesn't support polymorphism so we would need to use a `boost::variant<ComboAddress, DNSName, std::string>`, but sure. That would require creating the correct object beforehand:
```
> kvs:lookup(newDNSName("name.to.lookup."))
> kvs:lookup(newCA("192.0.2.1"))
```
instead of:
```
> kvs:lookup("name.to.lookup.")
> kvs:lookup("192.0.2.1")
```

I agree it would prevent surprises, though, so it's likely a good idea.